### PR TITLE
Add unique id for customer defined cdk functions

### PR DESCRIPTION
**Pull request by wchengru**
#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/providers/sam_function_provider.py [hunk 1]
```diff
@@ -235,12 +265,52 @@ def _convert_sam_function_resource(
             LOG.debug("Found Serverless function with name='%s' and ImageUri='%s'", name, imageuri)
 
         return SamFunctionProvider._build_function_configuration(
-            stack, name, codeuri, resource_properties, layers, inlinecode, imageuri, use_raw_codeuri
+            stack, function_id, name, codeuri, resource_properties, layers, inlinecode, imageuri, use_raw_codeuri
         )
 
+    @staticmethod
+    def _get_function_id(resource_properties: Dict, logical_id: str) -> str:
```
Comment (mildaniel) [inline samcli/lib/providers/sam_function_provider.py:272]: I believe the metadata generated by CDK is not in the properties, it is on the resource's root level. Can you double check this please?

Comment (wchengru) [reply samcli/lib/providers/sam_function_provider.py:272]: You are right, in the template it is. We pull it into the properties in the function provider: https://github.com/aws/aws-sam-cli/blob/ec9f5d89971b823fa01071c96b7de90ea5e5303a/samcli/lib/providers/sam_function_provider.py#L133

Comment (mildaniel) [reply samcli/lib/providers/sam_function_provider.py:272]: I see, thanks!

**Comment by mildaniel** [comment]
Per our conversation with Jacob, can you add an integration test for invoking with this new function identifier? I started writing some CDK invoke integration tests in [this PR](https://github.com/aws/aws-sam-cli/pull/3492) that you can build off of.

File: samcli/commands/local/lib/local_lambda.py [hunk 2]
```diff
@@ -256,7 +258,12 @@ def _make_env_vars(self, function: Function) -> EnvironmentVariables:
         else:
             # Standard format
             LOG.debug("Environment variables overrides data is standard format")
-            overrides = self.env_vars_values.get(name, None)
+            # Precedence: function_id -> logical_id -> full_path, customer can use any of them
```
Comment (jfuss) [inline samcli/commands/local/lib/local_lambda.py:261]: Why is this the right order? Shouldn't the previous one (logical_id) be the first to be backwards compatible?

File: samcli/lib/providers/provider.py [hunk 3]
```diff
@@ -29,6 +29,8 @@ class Function(NamedTuple):
     Named Tuple to representing the properties of a Lambda Function
     """
 
+    # Function id, can be Logical ID or any function identifier to define a function in specific IaC
```
Comment (jfuss) [inline samcli/lib/providers/provider.py:32]: What about the function name? `name` can be the Function Name or the logical id (reading the comment below)

File: samcli/lib/providers/sam_function_provider.py [hunk 4]
```diff
@@ -78,23 +78,52 @@ def get(self, name: str) -> Optional[Function]:
         if not name:
             raise ValueError("Function name is required")
 
-        # support lookup by full_path
+        get_f = None
```
Comment (jfuss) [inline samcli/lib/providers/sam_function_provider.py:81]: It's not clear what `get_f` is. Can you make a more descriptive variable name for this to improve readability?

File: samcli/lib/providers/sam_function_provider.py [hunk 5]
```diff
@@ -78,23 +78,52 @@ def get(self, name: str) -> Optional[Function]:
...
+        if not get_f:
+            # If function is not found by full path, search through all functions
+
+            found_fs = []
+
+            for f in self.get_all():
+                if name in (f.function_id, f.name, f.functionname):
+                    found_fs.append(f)
+
+            # If multiple functions are found, only return one of them
```
Comment (jfuss) [inline samcli/lib/providers/sam_function_provider.py:96]: Why is returning one of the the right UX? Shouldn't we only invoke if we are sure that is the function they want to invoke?

File: samcli/lib/providers/sam_function_provider.py [hunk 6]
```diff
@@ -78,23 +78,52 @@ def get(self, name: str) -> Optional[Function]:
...
+
+        return get_f
 
     def _deprecate_notification(self, runtime: Optional[str]) -> None:
         if runtime in self._deprecated_runtimes:
             message = (
                 f"WARNING: {runtime} is no longer supported by AWS Lambda, "
                 "please update to a newer supported runtime. SAM CLI "
-                f"will drop support for all deprecated runtimes {self._deprecated_runtimes} on May 1st. "
+                f"will drop support for all deprecated runtimes {self._deprecated_runtimes} on May 1st, 2020. "
```
Comment (jfuss) [inline samcli/lib/providers/sam_function_provider.py:126]: The linked Github issue is closed. Not in this update, but we should probably remove this as it doesn't look like we support these anymore.

File: samcli/lib/providers/sam_function_provider.py [hunk 7]
```diff
@@ -235,12 +265,52 @@ def _convert_sam_function_resource(
...
+        -------
+        str
+            The unique function id
+        """
+        resource_cdk_path = resource_properties.get("Metadata", {}).get("aws:cdk:path")
+
+        if not isinstance(resource_cdk_path, str) or not resource_cdk_path:
+            return logical_id
+
+        # aws:cdk:path metadata format: {stack_id}/{function_id}/Resource
```
Comment (jfuss) [inline samcli/lib/providers/sam_function_provider.py:296]: Are we sure this is the only format to `aws:cdk:path`? Maybe link out to docs?

**Review by jfuss — CHANGES_REQUESTED** [review]
Can we add integ tests for the changes to ensure this works as we expect it too?

File: samcli/lib/providers/sam_function_provider.py (hunk 7 above)
Comment (wchengru) [reply samcli/lib/providers/sam_function_provider.py]: Sure. I have added the cdk document about this in the comment.

File: samcli/lib/providers/sam_function_provider.py (hunk 6 above)
Comment (wchengru) [reply samcli/lib/providers/sam_function_provider.py]: Thanks. I removed this line, and only kept the warning, and the Github issue link.

File: samcli/lib/providers/sam_function_provider.py (hunk 5 above)
Comment (wchengru) [reply samcli/lib/providers/sam_function_provider.py:96]: Here the logic is kept the same as before for backward compatibility. I added a warning there because I noticed that in the nested stack case like this:
`StackA\FunctionA`
`StackA\NestedStackB\FunctionA`
The command `sam local invoke FunctionA` will only invoke one of them, and the order is also not promised. The change here is to (1) Make sure the invoked function is the same for every run. (2) Warn the user and provide the list of matched full paths.
I think if we stop the invoking, that will possible break the behaviour of some existing invocations usages.

File: samcli/lib/providers/sam_function_provider.py (hunk 4 above)
Comment (wchengru) [reply samcli/lib/providers/sam_function_provider.py]: Thanks, I have updated this.

File: samcli/lib/providers/provider.py (hunk 3 above)
Comment (wchengru) [reply samcli/lib/providers/provider.py:32]: Looks like the `name` is function name is always logical ID. When the `FunctionName` is not defined, the function name is the same as logical id. I updated that comment to make it more clear.
https://github.com/aws/aws-sam-cli/blob/ec9f5d89971b823fa01071c96b7de90ea5e5303a/samcli/lib/providers/sam_function_provider.py#L340

https://github.com/aws/aws-sam-cli/blob/ec9f5d89971b823fa01071c96b7de90ea5e5303a/samcli/lib/providers/sam_function_provider.py#L129

File: samcli/commands/local/lib/local_lambda.py (hunk 2 above)
Comment (wchengru) [reply samcli/commands/local/lib/local_lambda.py]: Thanks for pointing it out! I have fixed this order.

**Comment by wchengru** [comment]
> Per our conversation with Jacob, can you add an integration test for invoking with this new function identifier? I started writing some CDK invoke integration tests in [this PR](https://github.com/aws/aws-sam-cli/pull/3492) that you can build off of.



> Per our conversation with Jacob, can you add an integration test for invoking with this new function identifier? I started writing some CDK invoke integration tests in [this PR](https://github.com/aws/aws-sam-cli/pull/3492) that you can build off of.

I will add integration 

> Per our conversation with Jacob, can you add an integration test for invoking with this new function identifier? I started writing some CDK invoke integration tests in [this PR](https://github.com/aws/aws-sam-cli/pull/3492) that you can build off of.

Thanks @mildaniel for providing the integration test templates. I created integration tests based on your templates.

**Review by mildaniel — APPROVED** [review]

File: samcli/lib/providers/sam_function_provider.py (hunk 5 above)
Comment (jfuss) [reply samcli/lib/providers/sam_function_provider.py:96]: Seems reasonable. Thanks for explaining this. Could be worth an in comment code for future reference.

File: tests/integration/local/invoke/test_invoke_cdk_templates_with_function_id.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            # We don't actually use the build dir so we don't care if it's removed before another process finishes
+            shutil.rmtree(str(Path().joinpath(".aws-sam")))
+        except FileNotFoundError:
+            pass
+
+    @parameterized.expand([
+        ("StandardZipFunctionWithFunctionName", "ThisIsHelloWorldFunction", "LambdaWithFunctionName"),
+    ])
+    @pytest.mark.flaky(reruns=3)
+    def test_invoke_function_with_function_id(self, logical_id, function_name, function_id):
```
Comment (jfuss) [inline tests/integration/local/invoke/test_invoke_cdk_templates_with_function_id.py:49]: Should we have any integs tests around the precedence order or different edge cases we may encounter with templates?

Comment (wchengru) [reply tests/integration/local/invoke/test_invoke_cdk_templates_with_function_id.py:51]: I have added integration test cases for invoking functions in the nested stacks.
For the precedence on the env var overrides, the cases are unit tested:
https://github.com/aws/aws-sam-cli/pull/3498/files#diff-09b566488e9a422abe5c417dd6065f26119bd9bca9e9524e379c0534258a539fR223
